File: go/1464.go
package main

import "container/heap"

type Heap []int

// Len implements heap.Interface
func (h *Heap) Len() int {
	return len(*h)
}

// Less implements heap.Interface
func (h *Heap) Less(i int, j int) bool {
	return (*h)[i] >= (*h)[j]
}

// Swap implements heap.Interface
func (h *Heap) Swap(i int, j int) {
	(*h)[i], (*h)[j] = (*h)[j], (*h)[i]
}

// Pop implements heap.Interface
func (h *Heap) Pop() interface{} {
	ret := (*h)[len(*h)-1]
	*h = (*h)[:len(*h)-1]
	return ret
}

// Push implements heap.Interface
func (h *Heap) Push(x interface{}) {
	*h = append(*h, x.(int))
}

func maxProduct(nums []int) int {
	var h *Heap = &Heap{}
	heap.Init(h)
	for _, v := range nums {
		heap.Push(h, v)
	}

	// return (largest-1)*(2nd_largest-1)
	return (heap.Pop(h).(int) - 1) * (heap.Pop(h).(int) - 1)
}

File: cpp/1464.cpp
#include <iostream>
#include <queue>

using namespace std;

class Solution {
public:
  int maxProduct(vector<int> &nums) {
    int ret = 0;
    priority_queue<int> heap;
    for (int n : nums) {
      heap.push(n);
    }
    int x = heap.top();
    heap.pop();
    int y = heap.top();
    return (x-1)*(y-1);
  }
};
